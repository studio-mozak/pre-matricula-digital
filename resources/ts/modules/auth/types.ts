export interface Auth {
  level: number;
  name: string;
  schools: string[];
}
