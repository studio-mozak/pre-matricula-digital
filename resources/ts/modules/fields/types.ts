import { Nullable } from '@/types';

export type SetValues = (fields: Record<string, unknown>) => void;

export interface FieldPageOption {
  id?: string;
  name: Nullable<string>;
  weight: number;
}

export interface FieldPage {
  id: Nullable<string>;
  name: Nullable<string>;
  description: Nullable<string>;
  group: Nullable<'RESPONSIBLE' | 'STUDENT'>;
  type: Nullable<string>;
  internal: boolean;
  options: FieldPageOption[];
  deleteOptions?: string[];
  required?: boolean;
}

export interface ErrorResponse {
  message: string;
}
