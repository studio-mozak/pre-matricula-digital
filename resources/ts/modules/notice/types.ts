import { Nullable } from '@/types';

export interface Notice {
  id: Nullable<number>;
  text: Nullable<string>;
}

export interface NoticeSubmitResponse {
  errors: boolean;
  id?: number;
}
