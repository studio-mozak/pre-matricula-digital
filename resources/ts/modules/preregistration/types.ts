import {
  Address,
  ErrorResponse,
  Nullable,
  PaginatorInfo,
  Rules,
} from '@/types';
import { ProcessStageRestriction } from '@/modules/processes/types';

export interface PreRegistrationOverview {
  id: number;
  protocol: string;
  code: string;
  type: string;
  date: string;
  position: number;
  school: {
    id: number;
    name: string;
    area_code: string;
    phone: string;
  };
  grade: {
    id: number;
    name: string;
  };
  period: {
    id: number;
    name: string;
  };
  process?: {
    showPriorityProtocol: boolean;
  };
}

export interface PreRegistrationResponsibleField
  extends PreRegistrationResponsible {
  relationType: Nullable<string>;
  address: Address;
  useSecondAddress: boolean;
  secondAddress: Address;
}

export interface Match {
  id: number;
  initials: string;
  dateOfBirth: string;
  type: ProcessStageRestriction;
  registration: {
    year: number;
    school: {
      id: number;
      name: string;
    };
    grade: {
      id: number;
      name: string;
    };
  };
}

export interface WaitingList {
  period: string;
  school: string;
}

export interface PreRegistrationStudentField extends PreRegistrationStudent {
  grade: Nullable<string>;
  period: Nullable<string>;
  school: Nullable<string>;
  useSecondSchool: Nullable<boolean>;
  secondSchool: Nullable<string>;
  secondPeriod: Nullable<string>;
  match: Nullable<Match>;
  waitingList: WaitingList[];
  responsible_name?: string;
  responsible_cpf?: string;
  responsible_date_of_birth?: string;
  responsible_phone?: string;
}

export interface PreRegistrationList {
  id: string;
  type: string;
  protocol: string;
  status: string;
  student: {
    name: string;
    initials: string;
  };
  grade: {
    name: string;
  };
  period: {
    name: string;
  };
  school: {
    name: string;
  };
  schoolYear: {
    year: string;
  };
  position: number;
  waiting?: {
    id: string;
    type: string;
    status: string;
    position: number;
    student: {
      name: string;
    };
    period: {
      name: string;
    };
    school: {
      name: string;
    };
    schoolYear: {
      year: string;
    };
  };
  parent: {
    id: string;
    type: string;
    status: string;
    position: number;
    student: {
      name: string;
    };
    period: {
      name: string;
    };
    school: {
      name: string;
    };
    schoolYear: {
      year: string;
    };
  };
}

export interface StatProcesses {
  id?: string;
  accepted: number;
  rejected: number;
  total: number;
  vacancies: number;
  waiting: number;
}

export interface PreRegistrationFilter {
  template: string;
  protocol: string;
}

export interface Filter {
  first?: number;
  search: Nullable<string>;
  process: Nullable<string>;
  school: Nullable<string>;
  period?: Nullable<string>;
  grade: Nullable<string>;
  grades: Nullable<string[]>;
  type?: Nullable<string>;
  status?: Nullable<string>;
  sort: Nullable<string>;
  canAcceptInBatch: boolean;
  year: Nullable<number>;
  schools: Nullable<string[]>;
  processes: Nullable<string[]>;
}

export interface PreRegistrationStageProcessSchool {
  id: string;
  name: string;
  latitude: number;
  longitude: number;
}

export interface PreRegistrationStageProcessGrade {
  id: string;
  name: string;
  startBirth: string;
  endBirth: string;
}

export interface PreRegistrationStageProcessVacancy {
  school: number;
  period: number;
  grade: number;
  total: number;
  available: number;
}

export interface PreRegistrationStageProcessPeriod {
  id: string;
  name: string;
}

export interface PreRegistrationStageProcess {
  id: string;
  name: string;
  messageFooter: string;
  gradeAgeRangeLink: Nullable<string>;
  forceSuggestedGrade: boolean;
  showPriorityProtocol: boolean;
  schoolYear: {
    year: number;
  };
  grades: PreRegistrationStageProcessGrade[];
  periods: PreRegistrationStageProcessPeriod[];
  schools: PreRegistrationStageProcessSchool[];
  fields: ProcessField[];
  vacancies: PreRegistrationStageProcessVacancy[];
  allowResponsibleSelectMapAddress: boolean;
  blockIncompatibleAgeGroup: boolean;
  autoRejectByDays: boolean;
  autoRejectDays: Nullable<number>;
  onePerYear: boolean;
  waitingListLimit: number;
  minimumAge: Nullable<number>;
}

export interface PreRegistrationStage {
  id: string;
  renewalAtSameSchool: boolean;
  allowWaitingList: boolean;
  allowSearch: boolean;
  radius: number;
  type: string;
  status: string;
  restrictionType: ProcessStageRestriction;
  process: PreRegistrationStageProcess;
}

export interface PreRegistrationProcess {
  id: string;
  name: string;
  fields: ProcessField[];
  schoolYear: {
    year: number;
  };
  grades: {
    id: string;
    name: string;
  }[];
  periods: {
    id: string;
    name: string;
  }[];
  schools: {
    id: string;
    name: string;
    lat: string;
    lng: string;
  }[];
  vacancies: {
    school: string;
    period: string;
    grade: string;
    total: string;
    available: string;
  }[];
}

export interface ExternalPerson {
  id: number;
  name: string;
  gender: Nullable<number>;
  dateOfBirth: Nullable<string>;
  cpf: Nullable<string>;
  rg: Nullable<string>;
  birthCertificate: Nullable<string>;
  phone: Nullable<string>;
  mobile: Nullable<string>;
  email: Nullable<string>;
  address: Nullable<string>;
  number: Nullable<string>;
  complement: Nullable<string>;
  neighborhood: Nullable<string>;
  postalCode: Nullable<string>;
}

export interface PreRegistrationStudent {
  student_name: Nullable<string>;
  student_date_of_birth: string;
  student_cpf: string;
  student_rg: string;
  student_marital_status: number;
  student_place_of_birth: number | string;
  student_city_of_birth: string;
  student_birth_certificate: string;
  student_gender: number;
  student_email: string;
  student_phone: string;
  student_mobile: string;
}

export interface PreRegistrationResponsible {
  responsible_name: Nullable<string>;
  responsible_date_of_birth: string;
  responsible_cpf: string;
  responsible_rg: string;
  responsible_marital_status: number;
  responsible_place_of_birth: number | string;
  responsible_city_of_birth: string;
  responsible_gender: number;
  responsible_email: string;
  responsible_phone: string;
  responsible_mobile: string;
  responsible_address: {
    postalCode: string;
    address: string;
    number: string;
    neighborhood: string;
    city: string;
    complement: string;
    manualChangeLocation: boolean;
    lat: string;
    lng: string;
  }[];
}

export interface PreRegistration {
  id: number;
  date: string;
  protocol: string;
  status: string;
  type: string;
  observation: string;
  code?: string;
  grade: {
    id: string;
    name: string;
    course: {
      name: string;
    };
  };
  period: {
    id: string;
    name: string;
  };
  school: {
    id: string;
    name: string;
  };
  classroom: {
    id: string;
    name: string;
  };
  position: number;
  waiting: {
    id: string;
    protocol: string;
    school: {
      id: string;
      name: string;
    };
    position: number;
  };
  parent: {
    id: string;
    protocol: string;
    school: {
      id: string;
      name: string;
    };
    position: number;
  };
  others: {
    id: string;
    protocol: string;
    school: {
      id: string;
      name: string;
    };
    position: number;
  }[];
  student: PreRegistrationStudent;
  responsible: PreRegistrationResponsible;
  relationType: string;
  fields: {
    id: string;
    value: string;
    field: {
      id: string;
      name: string;
      internal: string;
      group: string;
    };
  }[];
  process: PreRegistrationProcess;
  inClassroom: Nullable<{
    id: string;
    name: string;
    period: {
      id: string;
      name: string;
    };
  }>;
  external: Nullable<ExternalPerson>;
}

export interface ProcessField {
  id: string;
  order: number;
  field: ProcessFieldField;
  required: string | number | boolean;
  weight: number;
}

export interface ProcessFieldField {
  id: string;
  name: string;
  description: Nullable<string>;
  group: string;
  internal:
    | 'responsible_rg'
    | 'student_rg'
    | 'responsible_name'
    | 'student_name'
    | 'responsible_date_of_birth'
    | 'student_date_of_birth';
  type:
    | 'CPF'
    | 'RG'
    | 'PHONE'
    | 'DATE'
    | 'TIME'
    | 'EMAIL'
    | 'CHECKBOX'
    | 'RADIO'
    | 'SELECT'
    | 'GENDER'
    | 'NUMBER'
    | 'MARITAL_STATUS'
    | 'CITY'
    | 'BIRTH_CERTIFICATE'
    | 'TEXT';
  options: {
    id: string;
    key: string;
    name: string;
    label: string;
    weight: number;
  }[];
}

export interface ParseFieldFromProcess extends ProcessField {
  id: string;
  key: string;
  rules: Rules;
  classes: {
    [key: string]: boolean;
  };
  type: string;
  label: string;
  filter:
    | ((value: string) => string)
    | ((value: string) => {
        key: string | Nullable<number>;
        label: string;
      });
}

export interface Fields {
  responsible: ParseFieldFromProcess[];
  student: ParseFieldFromProcess[];
}

export interface PreregistrationLoad {
  errors: boolean;
  stats: StatProcesses;
  preregistrations: PreRegistrationList[];
  paginator: PaginatorInfo;
}

export interface SchoolYear {
  year: number;
}

export interface ReportOptions {
  template: number;
  showStudentShortName: boolean;
  disregardStudentsIeducar: boolean;
}

export interface Period {
  id: number | string;
  name: string;
}

export interface Grade {
  id: string;
  key: string;
  label: string;
  name: string;
}

export interface Processes {
  id: string;
  key: string;
  label: string;
  name: string;
  schoolYear: SchoolYear;
  totalPreRegistrations: number;
}

export interface School {
  id: string;
  key: string;
  label: string;
  name: string;
}

export interface GetPreregistrations {
  grades: Grade[];
  periods: Period[];
  processes: Processes[];
  schools: School[];
}

export interface GetProcess {
  id: string;
  name: string;
  criteria: string;
  key: string;
  label: string;
  totalPreRegistrations: number;
  schoolYear: {
    year: number;
  };
  grades: Grade[];
  periods: Period[];
  schools: School[];
  fields: ProcessField[];
}

export interface PreregistrationSubmitInput {
  process: string;
  stage: string;
  type: string;
  grade: Nullable<string>;
  period: Nullable<string>;
  school: Nullable<string>;
  optionalSchool: Nullable<string>;
  optionalPeriod: Nullable<string>;
  relationType: Nullable<string>;
  address: Address;
  externalPerson: Nullable<number>;
  optionalAddress: Nullable<Address>;
  responsible: {
    field: string;
    value: Nullable<string>;
  }[];
  student: {
    field: string;
    value: Nullable<string>;
  }[];
  waitingList: WaitingList[];
}

export interface PreregistrationSubmit {
  errors?: ErrorResponse[];
  preregistrations: PreRegistrationOverview[];
}

export interface PreregistrationClassroom {
  key: number;
  label: string;
  period: {
    name: string;
  };
  available?: number;
}

export interface PreregistrationBatchResponse {
  id: string;
  student: {
    name: string;
  };
}

export interface PreregistrationBatchAccept {
  errors?: ErrorResponse[];
  acceptPreRegistrations: PreregistrationBatchResponse[];
}

export interface PreregistrationBatchReject {
  errors?: ErrorResponse[];
  rejectPreRegistrations: PreregistrationBatchResponse[];
}

export interface PreregistrationBatchSummon {
  errors?: ErrorResponse[];
  summonPreRegistrations: PreregistrationBatchResponse[];
}

export interface PreRegistrationBatchProps {
  filter: Filter;
  preregistrations: string[];
  step?: string;
  processYear: Nullable<number>;
  modelValue: boolean;
}
export interface StatProcessesProps {
  stats: StatProcesses;
}
