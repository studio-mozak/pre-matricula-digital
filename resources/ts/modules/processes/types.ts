import { ErrorResponse, ID, IdName, Nullable, Option } from '@/types';
import { ProcessField, ProcessFieldField } from '../preregistration/types';

export type ProcessType =
  | 'REGISTRATION_RENEWAL'
  | 'REGISTRATION'
  | 'WAITING_LIST';

export type ProcessStageRestriction =
  | 'NONE'
  | 'REGISTRATION_LAST_YEAR'
  | 'REGISTRATION_CURRENT_YEAR'
  | 'NO_REGISTRATION_CURRENT_YEAR'
  | 'NO_REGISTRATION_PERIOD_CURRENT_YEAR'
  | 'NEW_STUDENT'
  | 'PRE_REGISTRATION';

export type ProcessRejectType =
  | 'NO_REJECT'
  | 'REJECT_SAME_PROCESS'
  | 'REJECT_SAME_YEAR';

export interface SchoolYear {
  year: number;
}

export interface Stage {
  endAt: string;
  id: string;
  name: string;
  startAt: string;
  type: ProcessType;
  status: string;
  observation?: string;
  totalWaitingPreRegistrations?: number;
}

export interface Process {
  id: string;
  name: string;
  active: boolean;
  key: string;
  label: string;
  schoolYear: SchoolYear;
  stages: Stage[];
}

export interface Grade {
  name: string;
}

export interface Period {
  id?: string;
  name: string;
}

export interface School {
  name: string;
}

export interface Student {
  name: string;
}

export interface Responsible {
  name: string;
}

export interface Preregistration {
  id: string;
  parent: Nullable<string>;
  period: Period;
  position: number;
  protocol: string;
  school: School;
  schoolYear: Nullable<SchoolYear>;
  status: string;
  student: Student;
  responsible: Responsible;
  type: ProcessType;
  waiting: Nullable<Preregistration>;
}

export interface PreregistrationPaginator {
  count: number;
  currentPage: number;
  lastPage: number;
  perPage: number;
  total: number;
}

export interface ProcessCount {
  accepted: number;
  id: string;
  rejected: number;
  total: number;
  vacancies: number;
  waiting: number;
}

export interface ProcessCheck {
  id: string;
  name: string;
  fields: ProcessField[];
  grades: {
    endBirth: string;
    id: string;
    name: string;
    startBirth: string;
  }[];
  periods: Period[];
  schoolYear: SchoolYear;
  schools: {
    id: string;
    latitude: number;
    longitude: number;
    name: string;
  }[];
  stages: Stage[];
  vacancies: {
    available: number;
    grade: string;
    period: string;
    school: string;
    total: number;
  }[];
}

export interface ProcessCreate {
  active: boolean;
  forceSuggestedGrade: boolean;
  gradeAgeRangeLink: Nullable<string>;
  grades: string[];
  id: Nullable<string>;
  messageFooter: string;
  name: Nullable<string>;
  periods: string[];
  selectedSchools: boolean;
  schools: string[];
  schoolYear: Nullable<string>;
  showPriorityProtocol: boolean;
  allowResponsibleSelectMapAddress: boolean;
  blockIncompatibleAgeGroup: boolean;
  autoRejectByDays: boolean;
  autoRejectDays: Nullable<number>;
  waitingListLimit: number;
  rejectType: ProcessRejectType;
  minimumAge: Nullable<number>;
  onePerYear: boolean;
  showWaitingList: boolean;
  criteria: Nullable<string>;
  grouper: Nullable<string>;
}

export interface ProcessCreateResponse {
  active: boolean;
  forceSuggestedGrade: boolean;
  gradeAgeRangeLink: Nullable<string>;
  grades: ID[];
  id: Nullable<string>;
  messageFooter: string;
  name: Nullable<string>;
  periods: ID[];
  selectedSchools: boolean;
  schools: ID[];
  schoolYear: {
    id: string;
  };
  showPriorityProtocol: boolean;
  allowResponsibleSelectMapAddress: boolean;
  blockIncompatibleAgeGroup: boolean;
  autoRejectByDays: boolean;
  autoRejectDays: Nullable<number>;
  waitingListLimit: number;
  rejectType: ProcessRejectType;
  minimumAge: Nullable<number>;
  onePerYear: boolean;
  showWaitingList: boolean;
  criteria: Nullable<string>;
  grouper?: {
    id: string;
  };
}

export interface ProcessShow {
  fields: {
    field: {
      name: string;
      group: 'STUDENT' | 'RESPONSIBLE';
      type: string;
    };
    id: string;
    required: boolean;
    weight: number;
  }[];
  id: string;
  name: string;
  periods: {
    id: string;
    name: string;
  }[];
  schoolYear: {
    year: string;
  };
  stages: Stage[];
}

export interface ProcessStage {
  allowWaitingList: boolean;
  endAt: Nullable<string>;
  endHourAt: Nullable<string>;
  id: Nullable<string>;
  observation: Nullable<string>;
  radius: Nullable<number>;
  renewalAtSameSchool: boolean;
  startAt: Nullable<string>;
  startHourAt: Nullable<string>;
  type: string;
  restrictionType: ProcessStageRestriction;
  allowSearch: boolean;
}

export interface Course {
  id: string;
  name: string;
  grades: {
    id: string;
    name: string;
  }[];
}

export interface Field {
  priority: boolean;
  open: boolean;
  checked: string | number | boolean;
  weight: number;
  order: number;
  id: string;
  field?: ProcessFieldField;
  required: string | number | boolean;
  group?: 'STUDENT' | 'RESPONSIBLE';
  mandatory?: boolean;
  internal?: string;
  name?: string;
  options?: Option[];
  type?: string;
}

export interface Fields {
  field: string;
  order: number;
  required: string | number | boolean;
  weight: number;
}

export interface ProcessGetList {
  processes: Process[];
  years: Option[];
  status: Option[];
}

export interface ProcessPostAction {
  errors?: ErrorResponse[];
  process: ProcessShow;
}

export interface ProcessRejectInBatch {
  id: string;
  name: string;
  totalWaitingPreRegistrations: number;
}

export interface ProcessRejectInBatchResponse {
  errors?: ErrorResponse[];
  rejectInBatch: number;
}

export interface ProcessStageResponse {
  process: {
    id: string;
    name: string;
  };
  stages: ProcessStage[];
}

export interface ProcessPostStage {
  id: Nullable<string>;
  type: string;
  name: string;
  radius: number;
  startAt: string;
  endAt: string;
  observation: Nullable<string>;
  renewalAtSameSchool: string;
  allowWaitingList: string;
  restrictionType: number;
  allowSearch: boolean;
}

export interface ProcessFieldResponse {
  id: string;
  name: string;
  fields: {
    field: {
      id: string;
    };
    order: number;
    required: boolean;
    weight: number;
  }[];
}

export interface ProcessFieldList {
  process: ProcessFieldResponse;
  responsibleFields: Field[];
  studentFields: Field[];
}

export interface ProcessCreateCourses extends IdName {
  grades: {
    data: IdName[];
  };
}

export interface ProcessCreateCoursesResponse extends IdName {
  grades: IdName[];
}

export interface ProcessCreateListCourses {
  courses: ProcessCreateCoursesResponse[];
  periods: Period[];
  schoolYears: Option[];
  schools: Option[];
  groupers: Option[];
}

export interface ProcessGrouper {
  id: string;
  name: string;
  waitingListLimit: number;
  processes: Process[];
}
