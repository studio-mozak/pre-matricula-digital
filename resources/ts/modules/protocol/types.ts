import { ID, Nullable } from '@/types';
import {
  PreRegistrationOverview,
  PreRegistrationStageProcessGrade,
} from '@/modules/preregistration/types';

export interface FindProtocolPreRegistration {
  id: number;
  protocol: string;
  status: string;
  process: {
    schoolYear: {
      year: number;
    };
  };
  student: {
    initials: string;
    dateOfBirth: string;
  };
}

export interface ProtocolStatusPreRegistration {
  classroom: {
    grade: {
      name: string;
    };
    name: string;
    period: {
      name: string;
    };
  };
  id: string;
  observation: string;
  parent: Nullable<PreRegistrationOverview>;
  position?: number;
  process: {
    showPriorityProtocol: boolean;
    blockIncompatibleAgeGroup: boolean;
    forceSuggestedGrade: boolean;
    grades: PreRegistrationStageProcessGrade[];
  };
  school: {
    area_code: number;
    name: string;
    phone: string;
  };
  stage: {
    observation: string;
  };
  status: 'ACCEPTED' | 'WAITING' | 'REJECTED' | 'SUMMONED' | 'IN_CONFIRMATION';
  student: {
    dateOfBirth: string;
    initials: string;
  };
  type: string;
  waiting: Nullable<PreRegistrationOverview>;
}

export interface ProtocolStatusReturnToWait {
  errors: boolean;
  keepOnTheWaitingList: ID;
}
