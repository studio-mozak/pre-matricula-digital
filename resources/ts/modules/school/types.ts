import { SchoolYear, Stages } from '@/types';
export interface Grade {
  id: string;
  name: string;
}
export interface GradeFiltered {
  key: string;
  label: string;
}
export interface Processes {
  id: string;
  name: string;
  grades: Grade[];
  schoolYear: SchoolYear;
  stages: Stages[];
}
export interface SchoolInfo {
  id: string;
  name: string;
  lat: number;
  lng: number;
  area_code: number;
  phone: string;
}
export interface School extends SchoolInfo {
  title?: string;
  position: {
    lat: number;
    lng: number;
  };
}
export interface Vacancies {
  grade: string;
  period: string;
  process: string;
  school: string;
}

export interface ShowVacanciesReturn {
  vacancies: Vacancies[];
  schools: School[];
}
