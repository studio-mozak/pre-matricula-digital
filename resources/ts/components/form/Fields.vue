<template>
  <div class="row">
    <x-field
      v-for="(field, index) in modelFields"
      :key="index"
      v-model="modelData[field.key]"
      :container-class="field.classes"
      :rules="formatRules(field.rules)"
      :label="field.field.name"
      :description="field.field.description"
      :name="field.key"
      :type="field.type"
      :options="field.field.options"
      :errors="!!errors[field.key]"
    />
  </div>
</template>

<script lang="ts">
import { Option, Rules } from '@/types';
import { PropType, defineComponent } from 'vue';
import { ParseFieldFromProcess } from '@/modules/preregistration/types';
import XField from '@/components/x-form/XField.vue';
import { useVModel } from '@vueuse/core';

class FieldsFactory<T> {
  define() {
    return defineComponent({
      name: 'Fields',
      components: {
        XField,
      },
      props: {
        data: {
          type: Object as PropType<T>,
          default: () => ({}),
        },
        fields: {
          type: Array as PropType<ParseFieldFromProcess[]>,
          default: () => [],
        },
        errors: {
          type: Object as PropType<Rules>,
          default: () => ({}),
        },
        cities: {
          type: Array as PropType<Option[]>,
          default: () => [],
        },
      },
      setup(props) {
        const modelFields = useVModel(props, 'fields');
        const modelData = useVModel(props, 'data') as unknown as {
          [key: string]: string | string[] | boolean | number;
        };

        const formatRules = (rules: Rules) => {
          if (typeof rules === 'string') return rules;
          const keys = Object.keys(rules);
          const ruleStr: string[] = [];
          keys.forEach((key) => {
            if (typeof rules[key] === 'boolean') {
              if (rules[key]) {
                ruleStr.push(key);
              }
            } else {
              ruleStr.push(`${key}:${rules[key]}`);
            }
          });
          return ruleStr.join('|');
        };

        return {
          modelFields,
          modelData,
          formatRules,
        };
      },
    });
  }
}

const main = new FieldsFactory().define();
export function GenericFields<T>() {
  return main as ReturnType<FieldsFactory<T>['define']>;
}
export default main;
</script>
