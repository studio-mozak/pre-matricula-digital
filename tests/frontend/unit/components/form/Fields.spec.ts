import Fields, { GenericFields } from '@/components/form/Fields.vue';
import { Option, Rules } from '@/types';
import {
  ParseFieldFromProcess,
  PreRegistrationResponsibleField,
} from '@/modules/preregistration/types';
import { VueWrapper, flushPromises, mount } from '@vue/test-utils';
import { expect, vi } from 'vitest';

interface FieldsProps {
  data: PreRegistrationResponsibleField;
  fields: ParseFieldFromProcess[];
  errors: Rules;
  cities: Option[];
}

const propsMock: FieldsProps = {
  errors: {},
  cities: [],
  fields: [
    {
      id: 'string',
      key: 'string',
      rules: {},
      classes: {},
      type: 'TEXT',
      label: 'string',
      order: '0',
      field: {
        id: 'string',
        name: 'string',
        group: 'string',
        internal: 'responsible_name',
        type: 'TEXT',
        options: [],
      },
      required: true,
      weight: 0,
      filter: (value: string) => value,
    },
  ],
  data: {
    relationType: null,
    address: {
      postalCode: '',
      address: '',
      number: '',
      complement: '',
      neighborhood: '',
      lat: 0,
      lng: 0,
      city: '',
      cityIbgeCode: 0,
      stateAbbreviation: '',
    },
    useSecondAddress: false,
    secondAddress: {
      postalCode: '',
      address: '',
      number: '',
      complement: '',
      neighborhood: '',
      lat: 0,
      lng: 0,
      city: '',
      cityIbgeCode: 0,
      stateAbbreviation: '',
    },
    responsible_name: null,
  } as PreRegistrationResponsibleField,
};

const onMounted = async (props: FieldsProps) => {
  const wrapper = mount(GenericFields<PreRegistrationResponsibleField>(), {
    props,
    global: {
      stubs: ['x-field'],
    },
  });

  await wrapper.vm.$nextTick();

  await flushPromises();

  return wrapper;
};

describe('Fields', () => {
  let wrapper: VueWrapper<InstanceType<typeof Fields>>;

  beforeEach(async () => {
    wrapper = await onMounted(propsMock);

    await wrapper.vm.$nextTick();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('the component exists', () => {
    expect(wrapper).toBeTruthy();
  });

  test('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('one field must be rendered', () => {
    const fields = wrapper.findAll('x-field-stub');

    expect(fields.length).toBe(1);
  });

  test('zero fields must be rendered', async () => {
    wrapper = await onMounted({
      ...propsMock,
      fields: [],
    });

    const fields = wrapper.findAll('x-field-stub');

    expect(fields.length).toBe(0);
  });
});
