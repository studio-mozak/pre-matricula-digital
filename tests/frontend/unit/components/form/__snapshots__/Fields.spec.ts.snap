// Vitest Snapshot v1

exports[`Fields > should match snapshot 1`] = `
<div
  class="row"
>
  
  <x-field-stub
    allowclear="false"
    bind="[object Object]"
    containerclass="[object Object]"
    disabled="false"
    errors="false"
    label="string"
    loading="false"
    name="string"
    options=""
    rules=""
    searchable="false"
    type="TEXT"
  />
  
</div>
`;
